src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7667,7 +7670,7 @@ HRESULT ProfToEEInterfaceImpl::GetNonGCHeapBounds(ULONG cObjectRanges,
             ranges[segIdx].rangeStart = (ObjectID)firstObj;
 
             // Total size reserved for a segment
-            ranges[segIdx].rangeLengthReserved = (UINT_PTR)segments[segIdx]->m_Size;
+            ranges[segIdx].rangeLengthReserved = (UINT_PTR)segments[segIdx]->m_Size - sizeof(ObjHeader);
```

Comment:
segments[segIdx]->m_Size isn't the size of the reserved memory region?

Reply:
### Synopsis
This change is meant to be consistent with the inaccurate convention that already exists in `GetGenerationBounds`. 

### Root Cause Explained
The real problem is that `GetGenerationBounds` always report a number that is less than the actual reserved memory.

Here is how a segment/region is typically described in the GC.

```
heap_segment_reserved  <- the end address of the reservation
...
heap_segment_allocated <- the end of the last object in the segment/region
...
heap_segment_mem       <- where it points to the first object
...
unnamed                <- where the memory reservation actually starts
```

> As usual, higher means bigger address.

When we report the segments through `DiagDescrGenerations`, we reported

`heap_segment_mem`, `heap_segment_allocated` and `heap_segment_reserved` as `rangeStart`, `rangeEnd` and `rangeEndReserved` respectively.

Therefore when `GetGenerationBounds` reported the `rangeLengthReserved` as `rangeEndReserved - rangeStart`, it is not actually reporting the full memory reservation range, instead, the distance between `heap_segment_mem` and `unnamed` is omitted.

That length is variable depending on build flavor and runtime switches. It must at least allow for the object header to stay, but it also needs to fit some other GC needs. For example, the GC reserves a whole page there in case we are in segments with background GC enabled, or it could be `sizeof(aligned_plug_and_gap)` in case of regions to store relocation information.

For frozen segments created by `FrozenObjectHeap`, it doesn't need anything else but the object header, so the slack space between `unnamed` and `heap_segment_reserved` is exactly `sizeof(ObjHeader)`.

### Engineering trade-offs 
An alternative approach, obviously, is to fix `GetGenerationBounds`, and I chose not to. The mistake is actually consistently made in multiple places. Obviously, this also impacts the `GCGenerationBounds` event, but the mistake is also made through the DAC as well. Making fixes in all these places is a much bigger risk to take.

Yet another alternative is to ignore that. In which we risk inconsistency with what the DAC already reported.

The inaccuracies were probably always there since the beginning. It was initially reported as https://github.com/dotnet/diagnostics/issues/3482. The issue is currently closed. The fix is to adjust those values by the consumer of the DAC APIs. I expect if the profiler needs the exact number, they can do the same.

It is unfortunate that we chose to live with inaccuracies, but if we have to, let's at least be consistent.

Reply:
I see, m_size is the size but firstObject wasn't the true start of the range. Got it.